### PR TITLE
8323552: AbstractMemorySegmentImpl#mismatch returns -1 when comparing distinct areas of the same instance of MemorySegment

### PR DESCRIPTION
**Pull request by minborg**
Hi all,

This pull request contains a backport of commit [93579c29](https://github.com/openjdk/jdk/commit/93579c29e3ba60a8bc16d712e7ffc733f324a223) from the [openjdk/jdk](https://git.openjdk.org/jdk) repository.

The commit being backported was authored by Per Minborg on 25 Mar 2024 and was reviewed by Maurizio Cimadamore.

Thanks!

---------
### Progress
- [x] [JDK-8323552](https://bugs.openjdk.org/browse/JDK-8323552) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323552](https://bugs.openjdk.org/browse/JDK-8323552): AbstractMemorySegmentImpl#mismatch returns -1 when comparing distinct areas of the same instance of MemorySegment (**Bug** - P4 - Approved)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk22u.git pull/131/head:pull/131` \
`$ git checkout pull/131`

Update a local copy of the PR: \
`$ git checkout pull/131` \
`$ git pull https://git.openjdk.org/jdk22u.git pull/131/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 131`

View PR using the GUI difftool: \
`$ git pr show -t 131`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk22u/pull/131.diff">https://git.openjdk.org/jdk22u/pull/131.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk22u/pull/131#issuecomment-2042020778)

### CODE REVIEW

**Comment by minborg** [comment]
Please review this PR.

**Comment by minborg** [comment]
/approval request This is a clean backport of a bug discovered in 22

**Review by shipilev — APPROVED** [review]
This PR is clean, so it does not nominally requires review. You have maintainer approval for it, so it can go in.

**Pull request closed without merging**
